datafusion/core/src/physical_optimizer/enforcement.rs

```diff
@@ -0,0 +1,1739 @@
+                        parent_required,
+                        right.clone(),
+                        left.schema().fields().len(),
+                    )?,
+                    JoinType::RightSemi | JoinType::RightAnti => {
+                        adjust_input_keys_down_recursively(
+                            right.clone(),
+                            parent_required.clone(),
+                        )?
+                    }
+                    JoinType::Left
+                    | JoinType::LeftSemi
+                    | JoinType::LeftAnti
+                    | JoinType::Full => {
+                        adjust_input_keys_down_recursively(right.clone(), vec![])?
+                    }
+                };
+
+                Ok(Arc::new(HashJoinExec::try_new(
+                    new_left,
+                    new_right,
+                    on.clone(),
+                    filter.clone(),
+                    join_type,
+                    PartitionMode::CollectLeft,
+                    null_equals_null,
+                )?))
+            }
+        }
+    } else if let Some(CrossJoinExec { left, right, .. }) =
+        plan_any.downcast_ref::<CrossJoinExec>()
+    {
+        let new_left = adjust_input_keys_down_recursively(left.clone(), vec![])?;
+        let new_right = try_push_required_to_right(
+            parent_required,
+            right.clone(),
+            left.schema().fields().len(),
+        )?;
+        Ok(Arc::new(CrossJoinExec::try_new(new_left, new_right)?))
+    } else if let Some(SortMergeJoinExec {
+        left,
+        right,
+        on,
+        join_type,
+        sort_options,
+        null_equals_null,
+        ..
+    }) = plan_any.downcast_ref::<SortMergeJoinExec>()
+    {
+        let join_key_pairs = extract_join_keys(on);
+        if let Some((
+            JoinKeyPairs {
+                left_keys,
+                right_keys,
+            },
+            new_positions,
+        )) = try_reorder(
+            join_key_pairs.clone(),
+            parent_required,
+            &plan.equivalence_properties(),
+        ) {
+            let new_join_on = if !new_positions.is_empty() {
+                new_join_conditions(&left_keys, &right_keys)
+            } else {
+                on.clone()
+            };
+            let new_options = if !new_positions.is_empty() {
+                let mut new_sort_options = vec![];
+                for idx in 0..sort_options.len() {
+                    new_sort_options.push(sort_options[new_positions[idx]])
+                }
+                new_sort_options
+            } else {
+                sort_options.clone()
+            };
+
+            let new_left = adjust_input_keys_down_recursively(left.clone(), left_keys)?;
+            let new_right =
+                adjust_input_keys_down_recursively(right.clone(), right_keys)?;
+
+            Ok(Arc::new(SortMergeJoinExec::try_new(
```

Comment:
Since we changes the sort options here, will it influence upstream operators which may rely on it?

If this rule is called at the very beginning of the rule-based optimization process, this influence may be covered by other rules.